File: Programmers/[60060] 가사 검색/ONE/60060.java
import java.util.*;

class TrieNode {

    private int count;
    private boolean last;
    private Map<Character, TrieNode> childNodes = new HashMap<>();

    public Map<Character, TrieNode> getChildNodes() {
        return this.childNodes;
    }

    public int getCount() {
        return this.count;
    }

    public void increaseCount() {
        this.count++;
    }

    public boolean isLast() {
        return this.last;
    }

    public void setLast(boolean last) {
        this.last = last;
    }
}

class Trie {
    private TrieNode root;

    public Trie() {
        this.root = new TrieNode();
    }

    public void insert(String word) {
        int len = word.length();
        TrieNode node = this.root;

        for (int i = 0; i < len; i++) {
            node.increaseCount();
            node = node.getChildNodes().computeIfAbsent(word.charAt(i), c -> new TrieNode());
        }

        node.setLast(true);
    }

    public boolean contains(String word) {
        int len = word.length();
        TrieNode node = this.root;

        for (int i = 0; i < len; i++) {
            char c = word.charAt(i);
            TrieNode trieNode = node.getChildNodes().get(c);

            if(trieNode == null)
                return false;
            node = trieNode;
        }
        return node.isLast();
    }

    public int count(String keyword) {
        int len = keyword.length();
        TrieNode node = this.root;

        int count = 0;

        for (int i = 0; i < len; i++) {
            char c = keyword.charAt(i);
            if(c == '?')
                return node.getCount();
            TrieNode trieNode = node.getChildNodes().get(c);
            count = trieNode.getCount();
            node = trieNode;
        }
        return count;
    }
}

class Solution {
    public int[] solution(String[] words, String[] queries) {
        int n = queries.length;
        int[] answer = new int[n];
        Trie[] front = new Trie[10001];
        Trie[] back = new Trie[10001];

        for (String word : words) {
            int len = word.length();

            if (front[len] == null)
                front[len] = new Trie();
            front[len].insert(word);

            String reversed = new StringBuffer(word).reverse().toString();
            if (back[len] == null)
                back[len] = new Trie();
            back[len].insert(reversed);
        }

        for (int i = 0; i < n; i++) {
            String query = queries[i];
            int count = 0, len = query.length();

            try {
                if (query.charAt(0) == '?') {
                    query = new StringBuffer(query).reverse().toString();
                    count = back[len].count(query);
                } else
                    count = front[len].count(query);
            } catch (NullPointerException ignored) {
            }
            answer[i] = count;
        }
        return answer;
    }
}

File: Programmers/[60060] 가사 검색/JFe/60060.py
from collections import defaultdict
class Node:
    def __init__(self):
        self.count = 0
        self.flag = False
        self.child = defaultdict(Node)

class Trie:
    def __init__(self):
        self.root = Node()
        
    def insert(self, word):
        node = self.root
        node.count += 1
        for w in word:
            node = node.child[w]
            node.count += 1
        node.flag = True
    
    def search(self, word):
        node = self.root
        for w in word:
            if w not in node.child:
                return 0
            node = node.child[w]
        return node.count
    
def solution(words, queries):
    answer = []
    prefix = defaultdict(Trie)  # ?가 접두사로 올 때
    suffix = defaultdict(Trie)  # ?가 접미사로 올 때
    for word in words:
        # 단어 길이에 맞게 Trie 생성
        if len(word) not in prefix:
            prefix[len(word)] = Trie()
            suffix[len(word)] = Trie()
        prefix[len(word)].insert(word)
        suffix[len(word)].insert(word[::-1])
        
    for q in queries:
        cnt = q.count('?')
        # ?가 접두사로 올 때
        if q[0] == '?':
            answer.append(suffix[len(q)].search(q[cnt:][::-1]))
        # ?가 접미사로 올 때
        else:
            answer.append(prefix[len(q)].search(q[:len(q)-cnt]))
    return answer